cli/pa_analyze_command.go

```diff
@@ -0,0 +1,980 @@
+					actual:         int64(accountInfo.SubCnt),
+				},
+				{
+					description:    "Streams",
+					usageThreshold: 0.9,
+					limit:          accountInfo.Claim.Limits.Streams,
+					actual:         int64(len(r.GetAccountStreamNames(accountName))), // streams that belong to this account
+				},
+			}
+
+			// check all account limits
+			for _, accountLimit := range accountLimits {
+				// skip if limit is disabled or unlimited
+				if accountLimit.limit == 0 || accountLimit.limit == -1 {
+					continue
+				}
+
+				// calculate usage percentage
+				usage := 1.0 - (float64(accountLimit.limit-accountLimit.actual) / float64(accountLimit.limit))
+				if usage > accountLimit.usageThreshold {
+					examples.Addf("%s - %s: %d/%d", accountName, accountLimit.description, accountLimit.actual, accountLimit.limit)
+				}
+			}
+
+		}
+	}
+
+	if examples.Count() > 0 {
+		cmd.logIssue("Found accounts with high usage of limits")
+		cmd.logExamples(examples)
+		return SomeIssues, nil
+	}
+
+	return Pass, nil
+}
+
+// checkStreamLimits checks for streams that are close to their limits set in the stream config
+func (cmd *paAnalyzeCmd) checkStreamLimits(r *archive.Reader) (checkStatus, error) {
+
+	type streamLimit struct {
+		description    string
+		usageThreshold float64
+		limit          int64
+		actual         int64
+	}
+
+	typeTag := archive.TagStreamDetails()
+	accountNames := r.GetAccountNames()
+	examples := newCollectionOfExamples(cmd.examplesLimit)
+
+	for _, accountName := range accountNames {
+		accountTag := archive.TagAccount(accountName)
+		streamNames := r.GetAccountStreamNames(accountName)
+
+		for _, streamName := range streamNames {
+			streamTag := archive.TagStream(streamName)
+
+			serverNames := r.GetStreamServerNames(accountName, streamName)
+			for _, serverName := range serverNames {
+				serverTag := archive.TagServer(serverName)
+
+				var streamDetails server.StreamDetail
+
+				if err := r.Load(&streamDetails, serverTag, accountTag, streamTag, typeTag); errors.Is(err, archive.ErrNoMatches) {
+					cmd.logWarning("Artifact 'STREAM_DETAILS' is missing for stream %s in account %s", streamName, accountName)
+					continue
+				} else if err != nil {
+					return Skipped, fmt.Errorf("failed to load STREAM_DETAILS for stream %s in account %s: %w", streamName, accountName, err)
+				}
+
+				streamLimits := []streamLimit{
+					{
+						description:    "Messages",
+						usageThreshold: 0.95,
+						limit:          streamDetails.Config.MaxMsgs,
+						actual:         int64(streamDetails.State.Msgs),
+					},
+					{
+						description:    "Bytes",
+						usageThreshold: 0.95,
+						limit:          streamDetails.Config.MaxBytes,
+						actual:         int64(streamDetails.State.Bytes),
+					},
+					{
+						description:    "Consumers",
+						usageThreshold: 0.90,
+						limit:          int64(streamDetails.Config.MaxConsumers),
+						actual:         int64(streamDetails.State.Consumers),
```

Comment:
@jarretlavallee

Reply:
This one may trigger many false positives. When a stream hits max bytes or max messages, it will delete the oldest (or newest depending on config) message before inserting the next one. So many streams will be at the max bytes or max messages in a normal state. The consumer usage is valid and should work well.